src/qibo/backends/npmatrices.py

```diff
@@ -71,20 +75,25 @@ def M(self):  # pragma: no cover
         raise_error(NotImplementedError)
 
     def RX(self, theta):
+        theta = self._cast_parameter(theta)
         cos = self.np.cos(theta / 2.0) + 0j
         isin = -1j * self.np.sin(theta / 2.0)
         return self._cast([[cos, isin], [isin, cos]], dtype=self.dtype)
```

Comment:
```suggestion
        theta = self._cast_parameter(theta)
        cos = self.np.cos(theta / 2.0) + 0j
        isin = -1j * self.np.sin(theta / 2.0)
        return self.np.stack(([cos, isin, isin, cos])).reshape(2, 2)
```
I tested this locally and it seems to be working without the need of the convoluted workaround in `_cast`. I would say that updating all the parametrized gate this way might be better.

Reply:
This will create problems with flexible gates e.g. the new gate in https://github.com/qiboteam/qibo/pull/1373 (and also maybe `gates.Unitary`? I didn't test it but it seems so).

Reply:
What is the problem with the `GeneralizedRBS`? From what I see, you always build the matrix and cast it to the appropriate array type, thus `stack` should be fine as well, am I missing something? In general, anyways, this does not have to be applied to all the gates, you could always handle special cases with the current `_cast` in case it is needed (although at the moment no example comes to my mind). I tested the two alternative methods on `(2,2)` matrices and doing directly the `stack` seems to be 30 to 40% faster, clearly on a single gate call doesn't make any difference, but if you pile them up maybe....

Reply:
> What is the problem with the `GeneralizedRBS`? From what I see, you always build the matrix and cast it to the appropriate array type, thus `stack` should be fine as well, am I missing something? 

`GeneralizedRBS` and `Unitary` do not have a fixed matrix representation, so one cannot hardcode their flattened versions  like in your suggestion above.

> In general, anyways, this does not have to be applied to all the gates, you could always handle special cases with the current `_cast` in case it is needed (although at the moment no example comes to my mind). 

Sure, @Simone-Bordoni could add an exception just for `Unitary` right now (if actually needed) and then we handle `GeneralizedRBS` in #1373

Reply:
Ok nevermind, I tested numpy with the same direct `stack` and it seems to be way slower, not worth improving pytorch by a little bit while worsening numpy. At this point, either you rewrite the parametric gates in the `TorchMatrices` to maintain the small speed up at the cost of some code duplication, or you just live with a suboptimal pytorch matrix generation.

Reply:
I would just leave it as is. I know having two cast functions is not the prettiest. However, `_cast_parameter` is only needed because of `torch` and `tensorflow` and these two will be moved to `qiboml` soon, right?